editoast/src/views/infra/auto_fixes.rs

```diff
@@ -0,0 +1,191 @@
+use std::collections::HashSet;
+
+use crate::error::Result;
+use crate::generated_data::generate_infra_errors;
+use crate::infra_cache::InfraCache;
+use crate::models::{self, Infra, Retrieve};
+use crate::schema::operation::{DeleteOperation, Operation, OperationResult};
+use crate::schema::{
+    InfraError, InfraErrorType, OSRDIdentified, OSRDObject, ObjectRef, ObjectType,
+};
+use crate::DbPool;
+use actix_web::dev::HttpServiceFactory;
+use actix_web::get;
+use actix_web::web::{Data, Json as WebJson, Path};
+use chashmap::CHashMap;
+
+/// Return `/infra/<infra_id>/auto_fixes` routes
+pub fn routes() -> impl HttpServiceFactory {
+    list_auto_fixes
+}
+
+/// Return the list of suggestions to fix infra errors
+#[get("/auto_fixes")]
+async fn list_auto_fixes(
+    infra: Path<i64>,
+    infra_caches: Data<CHashMap<i64, InfraCache>>,
+    db_pool: Data<DbPool>,
+) -> Result<WebJson<Vec<Operation>>> {
+    let infra_id = infra.into_inner();
+    let infra = Infra::retrieve(db_pool.clone(), infra_id)
+        .await?
+        .ok_or(models::infra::InfraError::NotFound(infra_id))?;
+    let mut conn = db_pool.get().await?;
+
+    // accepting the early release of ReadGuard as it's anyway released when sending the suggestions (so before edit)
+    let mut infra_cache_clone = InfraCache::get_or_load(&mut conn, &infra_caches, &infra)
+        .await?
+        .clone();
+
+    let mut total_fix_iterations = 0;
+    let mut fixes = vec![];
+    let mut fixes_previous_len = 1; // force first fix
+    while fixes_previous_len != fixes.len() && total_fix_iterations < 5 {
+        total_fix_iterations += 1;
+        fixes_previous_len = fixes.len();
+        fixes.extend(fix_infra(&mut infra_cache_clone)?);
+    }
+
+    Ok(WebJson(fixes))
```

Comment:
I find it cleaner to write it like that:

```rust
    let mut fixes = vec![];
    for _ in 0..MAX_ITERATIONS { // MAX_ITERATIONS can be defined as const
        let new_fixes = fix_infra(&mut infra_cache_clone)?;
        if new_fixes.is_empty() {
            // We've fixed every possible error
            return Ok(WebJson(fixes));
        }
        fixes.extend(new_fixes);
    }
    
    // It feels safer to return a 500 error if we reach the maximum number of iteration
    // It preserves the property that reapplying an auto-fix should do nothing.
    Err(AutoFixesError::MaximumIterationReached.into())
```

What do you think?

Reply:
Much cleaner indeed, done in https://github.com/osrd-project/osrd/pull/5673/commits/7c7cd9b2be8e4f23eb926ae0377b640c54275e6e